### PR TITLE
[AOT] Annotate APIs which call `IServiceCollection.TryAddSingleton` for trim compatibility

### PR DESCRIPTION
**Pull request by vitek-karas**
Towards https://github.com/open-telemetry/opentelemetry-dotnet/issues/3429

`IServiceCollection.TryAddSingleton<T>` has trim annotation on `T`, this needs to be propagated to the caller to make it trim compatible.

This change just propagates the annotation to the signature of the method calling the `TryAddSingleton`. The trim tooling will make sure that the annotation is fulfilled by the caller of the OTel public methods with this change.

Note that these changes don't affect the outcome of the `EnsureAotCompatibility` test. This is because that test uses .NET 7 AOT compiler which has a bug and it doesn't report the warning it should have in these cases. Running the same test as a Trim compatibility test instead (`/p:PublishAot=false /p:PublishTrimmed=true`) will show warnings for each of the change in this PR. The .NET 8 version of the AOT compiler is fixed and will also report these warnings, so we need to fix these for both trimming and AOT compatibility.

I'd be open to adding a `EnsureTrimCompatibility` test, but it will make CI slower as it takes similar amount of time to the AOT compatibility test which is already causing CI slowdowns. @Yun-Ting has been discussing possibility of moving the AOT test into its own CI leg, adding the trim test there would be much less disruptive.

## Merge requirement checklist

* [ ] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/CONTRIBUTING.md) guidelines followed (nullable enabled, static analysis, etc.)
* [ ] Unit tests added/updated
* [ ] Appropriate `CHANGELOG.md` files updated for non-trivial changes
* [ ] Changes in public API reviewed (if applicable)

@Yun-Ting @eerhardt

### CODE REVIEW

File: src/OpenTelemetry.Api.ProviderBuilderExtensions/Logs/OpenTelemetryDependencyInjectionLoggerProviderBuilderExtensions.cs [hunk 1]
```diff
@@ -14,6 +14,9 @@
 // limitations under the License.
 // </copyright>
 
+#if NET6_0_OR_GREATER
```
Comment (eerhardt) [inline src/OpenTelemetry.Api.ProviderBuilderExtensions/Logs/OpenTelemetryDependencyInjectionLoggerProviderBuilderExtensions.cs:17]: We have UnconditionalSuppressMessageAttribute shim checked into the repo:

https://github.com/open-telemetry/opentelemetry-dotnet/blob/a808276efaf7e7a13420d993cea903cbff262996/src/Shared/Shims/UnconditionalSuppressMessageAttribute.cs#L30

Should we continue down that path here? Using both approaches feels wrong.

Comment (vitek-karas) [reply src/OpenTelemetry.Api.ProviderBuilderExtensions/Logs/OpenTelemetryDependencyInjectionLoggerProviderBuilderExtensions.cs:17]: The team decided to unify TFMs across the projects and include `net6.0` in all of them and then use ifdefs to only add the annotations to the `net6.0` versions.
So to make things consistent we should probably remove the shim for `UnconditionalSuppressMessageAttribute` and use ifdefs there as well.

@Yun-Ting as fyi

Comment (Yun-Ting) [reply src/OpenTelemetry.Api.ProviderBuilderExtensions/Logs/OpenTelemetryDependencyInjectionLoggerProviderBuilderExtensions.cs:17]: > The team decided to unify TFMs across the projects and include `net6.0` in all of them and then use ifdefs to only add the annotations to the `net6.0` versions. So to make things consistent we should probably remove the shim for `UnconditionalSuppressMessageAttribute` and use ifdefs there as well.
> 
I Agree. Please see the issue: https://github.com/open-telemetry/opentelemetry-dotnet/issues/4591

**Comment by utpilla** [comment]
> @Yun-Ting has been discussing possibility of moving the AOT test into its own CI leg, adding the trim test there would be much less disruptive.

+1

**Comment by vitek-karas** [comment]
I created https://github.com/open-telemetry/opentelemetry-dotnet/issues/4690 to track the work make a new CI leg for AOT/trim compatibility tests.

**Comment by utpilla** [comment]
> The trim tooling will make sure that the annotation is fulfilled by the caller of the OTel public methods with this change.

@vitek-karas Could you explain a bit more as to what this means for the users of our APIs?

**Comment by vitek-karas** [comment]
> > The trim tooling will make sure that the annotation is fulfilled by the caller of the OTel public methods with this change.

> @vitek-karas Could you explain a bit more as to what this means for the users of our APIs?

For users which don't trim/AOT their app there's no effect at all (the attribute is completely ignored by compiler and runtime).
For users which do trim/AOT the trimmer/AOT-compiler will check what is the type passed to the annotated `T`. If it's a specific type, the tool will preserve the necessary parts and there's nothing for the user to do - it just works. If it's another generic type (so somebody has a generic helper) they would get a warning and would need to propagate the annotation same way I did in this change. For example:

```C#
void RegisterInstrumentation<TInstrumentation>(LoggerProviderBuilder builder)
{
    builder.AddInstrumentation<TInstrumentation>(); // IL#### - the annotation on TInstrumentation doesn't match those required by AddInstrumentation
}
```

Fixed:

```C#
void RegisterInstrumentation<[DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicConstructor)] TInstrumentation>(LoggerProviderBuilder builder)
{
    builder.AddInstrumentation<TInstrumentation>(); // no warning
}
```

This is a standard behavior of many other APIs which are annotated with the `DynamicallyAccessedMembers` attribute, so it's likely that users which do trim/AOT would already be familiar with how it works.

**Comment by vitek-karas** [comment]
The test failure is a timed out AOT publish test. I ran it locally multiple times and it passes just fine. It also passed on Linux (in 1.5 minutes, on Windows it timed out after 4 minutes).

File: src/OpenTelemetry.Api.ProviderBuilderExtensions/Logs/OpenTelemetryDependencyInjectionLoggerProviderBuilderExtensions.cs (hunk 1 above)
Comment (Yun-Ting) [reply src/OpenTelemetry.Api.ProviderBuilderExtensions/Logs/OpenTelemetryDependencyInjectionLoggerProviderBuilderExtensions.cs:17]: https://github.com/open-telemetry/opentelemetry-dotnet/pull/4696

**Review by utpilla — APPROVED** [review]

**Review by reyang — APPROVED** [review]

**Pull request merged**